### PR TITLE
query: additional condition options

### PR DESCRIPTION
**Pull request by orfeasxyz**
A patch following up the issue I opened a while back https://github.com/labwc/labwc/issues/2126.

Other (more or less) related issues that I've come across:
https://github.com/labwc/labwc/issues/1837
https://github.com/labwc/labwc/issues/2023
https://github.com/labwc/labwc/issues/2029

List of options from Openbox and their implementation status (reference: http://openbox.org/help/Actions#If) :
- [x] shaded
- [x] maximized
- [x] maximizedhorizontal
- [x] maximizedvertical
- [x] iconified
- [x] focused
- [ ] urgent: There is no concept of urgent views yet.
- [x] undecorated: implemented differently (see below).
- [x] omnipresent
- [ ] activedesktop: Not implemented yet since it requires some extra changes to the way conditional actions are checked and executed. Currently all conditional actions are associated with one or more views, whereas "activedesktop" is associated with the active workspace. 
- [x] desktop
- [x] monitor

⭐ New in Labwc:
- [x] tiled ["up" | "right" | "down" | "left" | "center"]: succeeds if the current view is tiled in the respective direction.
- [x] tiled_region <region_name>: succeeds if the current view is tiled in the respective region.
- [x] decoration ["none" | "border" | "full"]: succeeds if the decoration mode of the current view matches with the given argument. Replaces openbox' \<undecorated\> query.
---

Some example uses:

A custom "ToggleMaximize" that UnSnaps the window back to its floating state
```xml
<keybind key="W-a">
  <action name="If">
    <query maximized="no"/>
    <then name.action="Maximize"/>
    <else name.action="UnSnap"/>
  </action>
</keybind>
```

Tile to top left region by tiling up and then left (for this to work for "down-left" as well, we need a way to express an `else if` statement).
```xml
<keybind key="W-Left">
  <action name="If">
    <query tiled="up"/>
    <then>
      <action name="SnapToRegion" region="up-left"/>
    </then>
    <else>
      <action name="SnapToEdge" direction="left"/>
    </else>
  </action>
</keybind>
```

Close all windows in other workspaces
```xml
<keybind key="W-S-q">
  <action name="ForEach">
    <query desktop="other"/>
    <then name.action="Close"/>
  </action>
</keybind>
```

Toggle between full and border decorations
```xml
<keybind key="W-d">
  <action name="If">
    <query decoration="full"/>
    <then>
      <action name="SetDecorations" decorations="border"/>
    </then>
    <else>
      <action name="SetDecorations" decorations="full"/>
    </else>
  </action>
</keybind>
```

### CODE REVIEW

**Comment by johanmalm** [comment]
Thanks for working on this. It's a nice addition.

Suggest we work towards merging after release of `0.8.1` (in a week's time).

There are some `./script/check` warnings to fix. All minor stuff.

```
src/config/rcxml.c:498: CHECK: Blank lines aren't necessary before a close brace '}'
src/view.c:154: CHECK: line length of 105 exceeds 100 columns
src/view.c:160: CHECK: line length of 101 exceeds 100 columns
src/view.c:178: CHECK: line length of 105 exceeds 100 columns
src/view.c:197: CHECK: line length of 108 exceeds 100 columns
src/view.c:199: CHECK: line length of 107 exceeds 100 columns
```

Would you mind updating `docs/labwc-actions.5.scd` with these additions?

The code looks good - except on construct that I'm not sure about. Will add review comment.

File: src/view.c [hunk 1]
```diff
@@ -68,19 +70,30 @@ struct view_query *
 view_query_create(void)
 {
 	struct view_query *query = znew(*query);
-	query->window_type = -1;
+	*query = (struct view_query) {
```
Comment (johanmalm) [inline src/view.c:73]: Can we do this against a heap allocated struct? Feels like `*query` will point to this static memory, which doesn't seem right if we have multiple queries and actually later change one of the members.

@ahesford @jlindgren90 - Does this construct make sense (and I'm just having a wobbly) or should it be changed?

**Review by johanmalm — COMMENTED** [review]
See note on initializing `query`

File: src/view.c (hunk 1 above)
Comment (orfeasxyz) [reply src/view.c:73]: The object on the right is stack allocated and copied into the memory pointed to by `query` (much like calling `memcpy`). It's also very likely the compiler will optimize it completely and write to the heap object directly.

**Review by ahesford — CHANGES_REQUESTED** [review]
I really don't like the reliance on magic `int` values to distinguish between unset and set true/false, especially because it leverages implementation details of the actual `bool` types we care about, when we should treat them as opaque, abstract types. I also don't like abusing bitwise operators to effect logical constructs (you're just inheriting this, but we might as well fix it right). I suggest using the already defined `enum three_state` in place of all the integers. I suggest the more explicit

```c
static bool
match_tristate(enum three_state desired, bool actual)
{
	switch (desired) {
	case LAB_STATE_ENABLED:
		retrun actual;
	case LAB_STATE_DISABLED:
		return !actual;
	default:
		return false;
	}
}
```

```c
if (match && query->some_tristate) {
	empty = false;
	match = match_tristate(query->some_tristate, view->actual_bool);
}
```

We don't need to combine the results of matches because we short-circuit tests after the first failure.

It might even be cleaner if we drop the `empty` and `match` pair altogether, redefine the match helper as

```c
static bool
match_tristate(enum three_state desired, bool actual)
{
	switch (desired) {
	case LAB_STATE_ENABLED:
		retrun actual ? LAB_STATE_ENABLED : LAB_STATE_DISABLED;
	case LAB_STATE_DISABLED:
		return actual ? LAB_STATE_DISABLED : LAB_STATE_ENABLED;
	default:
		return LAB_STATE_UNSPECIFIED;
	}
}
```

and then make `match` in the `view_matches_query` an `enum three_state`, allowing constructs like

```c
enum three_state match = LAB_STATE_UNSPECIFIED;
[...]
match = match_three_state(query->some_three_state, view->actual_bool);
if (match != LAB_STATE_UNSPECIFIED) {
	goto end_match;
}
[...]

end_match:
return match == LAB_STATE_ENABLED;
```
but this will require proper handling of the `match` variable in the string comparison blocks.

File: src/view.c (hunk 1 above)
Comment (johanmalm) [reply src/view.c:73]: Got it. Need to get more used to writing things that way :smile:

**Comment by orfeasxyz** [comment]
Thanks for pointing out the `struct three_state`, I will stick with the first implementation and refactor the function.

As a side note: I noticed that the <desktop> option is not working as intended. More specifically, one must specify the entire desktop name including the prefix (e.g. "Workspace 1"). I am thinking of using the `workspaces_find` function which means that the <desktop> query works as expected + we get a few extra options such as querying whether a view is on the last used workspace. A small problem however is that we cannot configure the wrap behavior for the "left" and "right" options so it either has to be hard-coded or implemented some other time.

**Comment by orfeasxyz** [comment]
@ahesford @johanmalm

I made some refactoring and added a few more options, if you have any suggestions LMK so I can finish everything up and write the docs.

**Comment by johanmalm** [comment]
> @ahesford @johanmalm
> 
> I made some refactoring and added a few more options, if you have any suggestions LMK so I can finish everything up and write the docs.

Look like you've inadvertently pushed some rogue commits. Please could you get rid of those.

@ahesford  - Would you be happy to continue the review / merge of this one.

**Comment by orfeasxyz** [comment]
> > @ahesford @johanmalm
> > I made some refactoring and added a few more options, if you have any suggestions LMK so I can finish everything up and write the docs.
> 
> Look like you've inadvertently pushed some rogue commits. Please could you get rid of those.
> 
> @ahesford - Would you be happy to continue the review / merge of this one.

Yeah it was an accident while I was trying to rebase, I will revert.

**Review by ahesford — COMMENTED** [review]
I pushed a commit that includes some doc cleanup and, more importantly, changes the interface for `maximized` queries from three separate booleans (which is cumbersom) to a multi-state `yes|no|horizontal|vertical`. Internally, we keep the three tri-state variables (with some renames to make them easier to distinguish), although we can refactor that later if desire strikes. I've also dropped all of your `wlr_log` calls in `view_matches_query`, because they were not universally applied to all tests anyway.

Please review these changes and, if you find them agreeable (and functional---I haven't actually run-tested them), we can squash and merge everything.

**Comment by ahesford** [comment]
Incidentally, if we were to completely eliminate the three tri-states for maximization, I would add a `VIEW_AXIS_INVALID = -1` to the definition of `enum view_axis` (making sure it doesn't break existing logic) and use that in the query struct.

**Pull request closed without merging**